fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -755,47 +777,90 @@ private static boolean unionHasMessageType(@Nonnull Descriptors.Descriptor union
+                    remapUnionField(newDescriptor, unionField);
+                }
             } else {
-                // Update the record type.
-                RecordTypeBuilder oldRecordType = recordTypes.get(descriptor.getName());
-                RecordTypeBuilder newRecordType = new RecordTypeBuilder(descriptor, oldRecordType);
-                recordTypes.put(newRecordType.getName(), newRecordType); // update the record type builder
-                unionFields.remove(oldRecordType.getDescriptor());
-                unionFields.put(descriptor, unionField);
+                if (oldDescriptor == null) {
+                    // New field and record type.
+                    RecordTypeBuilder recordType = processRecordType(unionField, processExtensionOptions);
+                    if (recordType.getSinceVersion() != null && recordType.getSinceVersion() != version) {
+                        throw new MetaDataException(String.format("Record type version (%d) does not match meta-data version (%d)",
+                                recordType.getSinceVersion(), version));
+                    } else {
+                        recordType.setSinceVersion(version);
+                    }
+                    unionFields.put(newDescriptor, unionField);
+                } else {
+                    // Create a new record type based off of the old one.
+                    RecordTypeBuilder oldRecordType = oldRecordTypes.get(oldDescriptor.getName());
+                    RecordTypeBuilder newRecordType = new RecordTypeBuilder(newDescriptor, oldRecordType);
+                    recordTypes.put(newRecordType.getName(), newRecordType); // update the record type builder
+                    unionFields.put(newDescriptor, unionField);
+                }
             }
         }
     }
 
-    private void fillUnionFields(@Nonnull Descriptors.Descriptor union, @Nullable Descriptors.Descriptor localUnionDescriptor, boolean processExtensionOptions) {
-        for (Descriptors.FieldDescriptor unionField : union.getFields()) {
-            Descriptors.FieldDescriptor adjustedUnionField = adjustUnionField(unionField, localUnionDescriptor);
-            Descriptors.Descriptor descriptor = adjustedUnionField.getMessageType();
+    @Nullable
+    private Descriptors.Descriptor findOldDescriptor(@Nonnull Descriptors.FieldDescriptor newUnionField, @Nonnull Descriptors.Descriptor newDescriptor, @Nonnull Descriptors.Descriptor newUnion) {
+        // If there is a corresponding field in the old union, use that field.
+        Descriptors.FieldDescriptor correspondingField = unionDescriptor.findFieldByNumber(newUnionField.getNumber());
+        if (correspondingField != null) {
+            return correspondingField.getMessageType();
+        }
+        // Look for a field in the new union of the same type as this one and look for a matching field in the old union
+        for (Descriptors.FieldDescriptor fieldDescriptor : newUnion.getFields()) {
```

Comment:
Thanks for the explanation. Right, I see that. This seems to me like a rename + addition, rather than just a rename. And, in my opinion, dropping the loop must work because of the `remapUnionField` logic. You return `null`, it considers it a new field and the last field of the same type takes precedence. I have no strong objection on this, but thinking thru it again, I still believe it's not necessary and may become a venue of bugs.

Reply:
I had tried to accomplish this before, but in the most recent commit, I was able to construct a test that fails if `findOldDescriptor` returns `null` here.

Yeah, renaming the record type is neither necessary nor sufficient to trigger the exceptional case; it all has to do with updates to the union descriptor. And in particular, because we iterate through the union fields in the order in which they are returned by `union.getFields()` which (AFAICT) is the same as file declaration order, it requires that the order of the fields is also changed. Something like:

```proto
message RecordTypeUnion {
   optional RecordOne _RecordOne = 1;
   optional RecordTwo _RecordTwo = 2;
}
```

To:

```proto
message RecordTypeUnion {
   optional RecordOne _NewRecordOne = 3;
   optional RecordOne _OldRecordOne = 1;
   optional RecordTwo _RecordTwo = 2;
}
```

Then `_NewRecordOne` is the first thing processed, `findOldDescriptor` returns `null`, so it creates a new record type for `RecordOne`. Then `_OldRecordOne` is processed, it *does* have an old descriptor, but then it will see it as just a separate field associated with `RecordOne` and update the union field. This can lead to the `since_version` being wrong on the record (among other problems).

In looking into this, I discovered #620, and I included a fix for it in the most recent commit. In theory, we should have an additional test for backwards compatibility. (I made a claim in #620 that all on-disk data, even for affected meta-data, would be compatible with and without the change, so there was no need for, say, a format version change.) But I also wanted to get this onto this branch so that people could see it.

Reply:
I see, thanks for getting to the bottom of this and for the tests. The changes look good to me.